bach/bach/expression.py

```
@@ -282,18 +299,49 @@ def has_windowed_aggregate_function(self) -> bool:
             d.has_windowed_aggregate_function for d in self.data if isinstance(d, Expression)
         )
 
+    @property
+    def has_table_column_references(self) -> bool:
+        """
+        True iff we are a TableColumnReference, or there is at least one in this Expression.
+        """
+        return any(
+            isinstance(token, TableColumnReferenceToken) for token in self.get_all_tokens()
+        )
+
     def resolve_column_references(self, dialect: Dialect, table_name: Optional[str]) -> 'Expression':
         """ resolve the table name aliases for all columns in this expression """
         result: List[Union[ExpressionToken, Expression]] = []
         for data_item in self.data:
             if isinstance(data_item, Expression):
                 result.append(data_item.resolve_column_references(dialect, table_name))
             elif isinstance(data_item, ColumnReferenceToken):
-                result.append(data_item.resolve(dialect, table_name))
+                result.append(data_item.resolve(table_name))
             else:
                 result.append(data_item)
         return self.__class__(result)
 
+    def remove_table_column_references(self) -> Tuple[str, str, 'Expression']:
+        """
+        removes all table references from this expression.
+        Returns first table_name and column_name found and a new expression without table column references
```

Should it not return as list of (table_name, column_name) tuples? The first one is quite random :)

---

I thought about returning a mapping between table_names and list of columns, buuuuuut make things more complicated and in theory we shouldn't be allowing multiple table refs in an expression.

---

also, usually the first column found in the expression is the name of the series. It's "hacky" but prefer to call it an heuristic ;)

---

is the "in theory" guarded aka do we raise when that happens? Maybe we should check in this method that it's actually u
unique or raise, that'd be nice?

---

nope, :thinking: the only place where we set the table alias is in "merge" and it's hardcoded. Not sure, in that case probably we will need to "preprocess" the expressions in the SqlModel

---

but will raise it here :P

---

I don't think it makes much sense to raise the exception only here. This is an okay second place to do it. But I think that if we say that `"we shouldn't be allowing multiple table refs in an expression"`, then we should raise that exception when we create such an illegal `Expression`.

---

Also, in a merge could we not have an expression that references two different tables, or am I misunderstanding that?

---

I do agree that returning a list or map of table-names wouldn't be very useful for code that calls this probably.